test/e2e/upgradepath_test.go

```diff
@@ -0,0 +1,234 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	name := "operator-upgrade"
+
+	// Create Alertmanager, Prometheus, Thanosruler services with selector labels promised by Prometheus Operator
+	alertmanagerService := v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("alertmanager-%s", name),
+			Labels: map[string]string{
+				"group": name,
+			},
+		},
+		Spec: v1.ServiceSpec{
+			Type: v1.ServiceTypeClusterIP,
+			Ports: []v1.ServicePort{
+				{
+					Name:       "web",
+					Port:       9093,
+					TargetPort: intstr.FromString("web"),
+				},
+			},
+			Selector: map[string]string{
+				"app.kubernetes.io/name":       "alertmanager",
+				"app.kubernetes.io/managed-by": "prometheus-operator",
+				"app.kubernetes.io/instance":   name,
+				"alertmanager":                 name,
+			},
+		},
+	}
+
+	prometheusService := v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("prometheus-%s", name),
+			Labels: map[string]string{
+				"group": name,
+			},
+		},
+		Spec: v1.ServiceSpec{
+			Type: v1.ServiceTypeClusterIP,
+			Ports: []v1.ServicePort{
+				{
+					Name:       "web",
+					Port:       9090,
+					TargetPort: intstr.FromString("web"),
+				},
+			},
+			Selector: map[string]string{
+				"app.kubernetes.io/name":       "prometheus",
+				"app.kubernetes.io/managed-by": "prometheus-operator",
+				"app.kubernetes.io/instance":   name,
+				"prometheus":                   name,
+				"operator.prometheus.io/shard": "0",
+				"operator.prometheus.io/name":  name,
+			},
+		},
+	}
+
+	thanosRulerService := v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("thanos-ruler-%s", name),
+			Labels: map[string]string{
+				"group": name,
+			},
+		},
+		Spec: v1.ServiceSpec{
+			Type: v1.ServiceTypeClusterIP,
+			Ports: []v1.ServicePort{
+				{
+					Name:       "web",
+					Port:       9090,
+					TargetPort: intstr.FromString("web"),
+				},
+			},
+			Selector: map[string]string{
+				"app.kubernetes.io/name":       "thanos-ruler",
+				"app.kubernetes.io/managed-by": "prometheus-operator",
+				"app.kubernetes.io/instance":   name,
+				"thanos-ruler":                 name,
+			},
+		},
+	}
+
+	alertmanager := framework.MakeBasicAlertmanager(name, 1)
+	_, err = framework.CreateAlertmanagerAndWaitUntilReady(context.Background(), ns, alertmanager)
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = framework.CreateOrUpdateServiceAndWaitUntilReady(context.Background(), ns, &alertmanagerService)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	framework.SetupPrometheusRBAC(context.Background(), t, testCtx, ns)
+	prometheus := framework.MakeBasicPrometheus(ns, name, name, 1)
+
+	// Method to wait for prometheus ready changed since v0.56.0.
+	// Hence we need to use the old way to check if prometheus ready for operator v0.55.x
+	// TODO(heylongdacoder): this condition checking can be removed when the latest operator version reach 0.57.0
+	if currentSemVer.LT(semver.MustParse("0.57.0")) {
```

Comment:
@heylongdacoder .57 was release this week, so I guess we can simplify this already? :)